hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/RetriableFileCopyCommand.java

```diff
@@ -197,14 +200,25 @@ private long copyToFile(Path targetPath, FileSystem targetFS,
           targetFS, targetPath);
       final long blockSize = getBlockSize(fileAttributes, source,
           targetFS, targetPath);
+      EnumSet<CreateFlag> createFlags =
+          EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE);
+      if (noLocalWrite) {
+        createFlags.add(CreateFlag.NO_LOCAL_WRITE);
+      }
       FSDataOutputStream out = targetFS.create(targetPath, permission,
-          EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE),
-          copyBufferSize, repl, blockSize, context,
+          createFlags, copyBufferSize, repl, blockSize, context,
           getChecksumOpt(fileAttributes, sourceChecksum));
       outStream = new BufferedOutputStream(out);
     } else {
-      outStream = new BufferedOutputStream(targetFS.append(targetPath,
-          copyBufferSize));
+      if (targetFS instanceof DistributedFileSystem && noLocalWrite) {
+        outStream = new BufferedOutputStream(
+            ((DistributedFileSystem) targetFS).append(targetPath,
```

Comment:
you need to do an explicit cast here? Ugly. And it will mean we need hadoop-hdfs on the classpath, always, which object stores may not have. We'd be better off (carefully) pulling that HDFS method up